### ISSUE TITLE
Missing copy protocol

### ISSUE THREAD

**Issue by jimy-byerley** [Zuzu-Typ/PyGLM#40]
Hi, first I have to congratulate you for this so conveinient library !
I tried the following:
`copy.deepcopy(glm.vec3(1))`
`copy.copy(glm.vec3(1))`

in both cases, I get the following error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/copy.py", line 96, in copy
    rv = reductor(4)
TypeError: can't pickle glm.vec3 objects
```

I think that for wrapped objects it's done using the `__reduce__` or the `__deepcopy__` method.
I use these vectors and matrices in classes to represent points of primitives, I need to copy these Primitive objects.
Can you please implement the copy protocol for the glm objects ?

**Comment by Zuzu-Typ** [comment]
Hi there jimy,

I've looked into the `copy` module and found that it's incompatible to PyGLM, as it tries to convert it into a byte string directly from memory, where in some cases there may be a pointer, of which the contents wouldn't be copied. There is a way to turn a PyGLM type into a byte-string though, as it supports the buffer protocol (`vec3_as_bytes = bytes(glm.vec3(1,2,3))`).

However, if what you're trying to accomplish is getting a non-reference copy of a `vec3`, there is a much simpler way of doing that.
`b = glm.vec3(a)`
This is what's known as a copy-constructor.

If you don't know whether it's a `vec3` or `ivec4` or `mat3` or whatever, you can simply use
`b = type(a)(a)` or `b = a.__class__(a)`.

Cheers,
--Zuzu_Typ--

**Comment by jimy-byerley** [comment]
> it tries to convert it into a byte string directly from memory, where in some cases there may be a pointer, of which the contents wouldn't be copied.

I see, but I wonder if the method __deepcopy__ couldn't be implemented for PyGLM types, __deepcopy__ is called whenever the object is in a complex object tree and juste have to return an object like him, in the case of PyGLM it would only need to call the copy constructor.

I think it's important to implement the copy protocol, because that allows to copy a complex data structure without the need to copy each of its elements individually by knowing if each object comes from PyGLM or from an other module.

**Comment by Zuzu-Typ** [comment]
Okay, I'll look into it when I have the time.
Until then, please try using this workaround:
```Python
class copyvec3(glm.vec3):
	def __copy__(self):
		return copyvec3(self)
	def __deepcopy__(self, memo):
		copy = copyvec3(self)
		memo[id(self)] = copy
		return copy
```
Alternatively you could also try using generator functions or your own implementation of the deepcopy method.

Thank you for the suggestion.

Cheers,
--Zuzu_Typ--

**Comment by Zuzu-Typ** [comment]
Should've been fixed by c10a030a1e15912944d650cd0d5b548022589ca8.
